### ISSUE TITLE
Rule/Perf: Reduce allocations in no-floating-promises

### ISSUE THREAD

**Issue by DanielRosenwasser** [typescript-eslint/typescript-eslint#12638]
### AI Assistance

- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Suggestion

In no-floating-promises, every `ExpressionStatement` invokes `isUnhandledPromise`, which recursively walks the tree. As it does, it reports information in the form of an object. These objects contain up to 3 `boolean` properties which are not always set.

```ts
    function isUnhandledPromise(
      checker: ts.TypeChecker,
      node: TSESTree.Node,
    ): {
      isUnhandled: boolean;
      nonFunctionHandler?: boolean;
      promiseArray?: boolean;
    }
```

I have not measured the cost yet, but instead of returning a new object with varying shapes, I figured these could be represented as bit flags on a small int.
